src/oskbd/macos.rs

```diff
@@ -0,0 +1,230 @@
+                None
+            }
+        })
+        .collect()
+}
+
+impl TryFrom<InputEvent> for KeyEvent {
+    type Error = ();
+
+    fn try_from(item: InputEvent) -> Result<Self, Self::Error> {
+        if let Ok(oscode) = OsCode::try_from(PageCode {
+            page: item.page,
+            code: item.code,
+        }) {
+            Ok(KeyEvent {
+                code: oscode,
+                value: if item.value == 1 {
+                    KeyValue::Press
+                } else {
+                    KeyValue::Release
+                },
+            })
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl TryFrom<KeyEvent> for InputEvent {
+    type Error = ();
+
+    fn try_from(item: KeyEvent) -> Result<Self, Self::Error> {
+        if let Ok(pagecode) = PageCode::try_from(item.code) {
+            let val = match item.value {
+                KeyValue::Press => 1,
+                _ => 0,
+            };
+            Ok(InputEvent {
+                value: val,
+                page: pagecode.page,
+                code: pagecode.code,
+            })
+        } else {
+            Err(())
+        }
+    }
+}
+
+pub struct KbdOut {}
+
+impl KbdOut {
+    pub fn new() -> Result<Self, io::Error> {
+        Ok(KbdOut {})
+    }
+
+    pub fn write(&mut self, event: InputEvent) -> Result<(), io::Error> {
+        let mut devent = event.into();
+        log::debug!("Attempting to write {event:?} {devent:?}");
+        let _sent = send_key(&mut devent);
+        Ok(())
+    }
+
+    pub fn write_key(&mut self, key: OsCode, value: KeyValue) -> Result<(), io::Error> {
+        if let Ok(event) = InputEvent::try_from(KeyEvent { value, code: key }) {
+            self.write(event)
+        } else {
+            log::debug!("couldn't write unrecognized {key:?}");
+            Err(io::Error::new(
+                io::ErrorKind::NotFound,
+                "OsCode not recognized!",
+            ))
+        }
+    }
+
+    pub fn write_code(&mut self, code: u32, value: KeyValue) -> Result<(), io::Error> {
+        if let Ok(event) = InputEvent::try_from(KeyEvent {
+            value,
+            code: OsCode::from_u16(code as u16).unwrap(),
+        }) {
+            self.write(event)
+        } else {
+            log::debug!("couldn't write unrecognized OsCode {code}");
+            Err(io::Error::new(
+                io::ErrorKind::NotFound,
+                "OsCode not recognized!",
+            ))
+        }
+    }
+
+    pub fn press_key(&mut self, key: OsCode) -> Result<(), io::Error> {
+        self.write_key(key, KeyValue::Press)
+    }
+
+    pub fn release_key(&mut self, key: OsCode) -> Result<(), io::Error> {
+        self.write_key(key, KeyValue::Release)
+    }
+
+    /// Send using C-S-u + <unicode hex number> + spc
+    pub fn send_unicode(&mut self, c: char) -> Result<(), io::Error> {
+        log::debug!("sending unicode {c}");
+        todo!();
```

Comment:
It's also fine to not implement this yet - maybe macos doesn't even have any unicode output solution?

If left unimplemented, I wonder if it's better to have a `log::error!` call rather than panic with a `todo`... I don't feel too strongly either way for now.

Reply:
Yeah, I'm not sure if Macos has that yet. I am quite new to the Mac ecosystem and IOKit but I can look into it.
For now I changed it to `log::error!` as suggested.